File: JPA/studyolle/Backend/src/main/java/me/kzv/olle/web/account/AccountService.java
package me.kzv.olle.web.account;

import lombok.RequiredArgsConstructor;
import me.kzv.olle.web.account.dto.SignUpRequestDto;
import org.springframework.mail.SimpleMailMessage;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class AccountService {
    private final AccountRepository accountRepository;
    private final JavaMailSender javaMailSender;
    private final PasswordEncoder passwordEncoder;


    @Transactional
    public Account processNewAccount(SignUpRequestDto signUpForm) {
        Account newAccount = saveNewAccount(signUpForm); //같은 transaction 안이어야 account 객체가 변경되어도 persistence 가 save 된다.
        newAccount.generateEmailCheckToken();
        sendSignUpConfirmEmail(newAccount);

        return newAccount;
    }

    private Account saveNewAccount(SignUpRequestDto signUpForm) {
        Account account = Account.builder()
                .email(signUpForm.getEmail())
                .nickname(signUpForm.getNickname())
                .password(passwordEncoder.encode(signUpForm.getPassword())) // TODO encoding
                .studyCreateByWeb(true)
                .studyEnrollmentResultByWeb(true)
                .studyUpdatedByWeb(true)
                .build();

        Account newAccount = accountRepository.save(account);
        return newAccount;
    }

    private void sendSignUpConfirmEmail(Account newAccount) {
        SimpleMailMessage mailMessage = new SimpleMailMessage();
        mailMessage.setSubject("스터디 올래, 회원 가입 인증"); // 제목
        mailMessage.setText("/check-email-token?token=" + newAccount.getEmailCheckedToken() + "&email=" + newAccount.getEmail()); // 본문

        javaMailSender.send(mailMessage);
    }

    public void login(Account account) {
        // 야메
//        UsernamePasswordAuthenticationToken token = new UsernamePasswordAuthenticationToken(account.getNickname(), account.getPassword(), List.of(new SimpleGrantedAuthority("ROLE_USER")));
//        SecurityContext context = SecurityContextHolder.getContext();
//        context.setAuthentication(token);

        // 정석
//        UsernamePasswordAuthenticationToken token = new UsernamePasswordAuthenticationToken(username, password);
//        Authentication authentication = authenticationManager.getContext();
//        SecurityContext context = SecurityContextHolder.getContext();
//        context.setAuthentication(token);


    }
}

File: toy/studyolle/src/main/kotlin/me/kzv/study/account/AccountService.kt
package me.kzv.study.account

import me.kzv.study.domain.Account
import org.springframework.mail.SimpleMailMessage
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.stereotype.Service

@Service
class AccountService(
    private val accountRepository: AccountRepository,
    private val javaMailSender: JavaMailSender,
) {

    fun processNewAccount(signUpForm: SignupForm) {
        val newAccount = saveNewAccount(signUpForm)
        newAccount.generateEmailCheckToken()
        sendSignUpConfirmEmail(newAccount)
    }

    private fun saveNewAccount(signUpForm: SignupForm): Account {
        val account = Account(
            email = signUpForm.email,
            nickname = signUpForm.nickname,
            password = signUpForm.password,
        )

        return accountRepository.save(account)
    }

    private fun sendSignUpConfirmEmail(newAccount: Account) {
        val mailMessage = SimpleMailMessage()
        mailMessage.setSubject("스터디올래, 회원 가입 인증") // 제목
        mailMessage.setText("/check-email-token?token=${newAccount.emailCheckToken}&email=${newAccount.email}")// 본문

        javaMailSender.send(mailMessage)
    }
}